File: protocol/python/bridge.py
import asyncio
import json
import sys
import subprocess
from typing import Dict, Any, Optional, Callable
from pathlib import Path
import signal

from .types import (
    ToolContext,
    ToolResult,
    JsonRpcRequest,
    JsonRpcResponse,
    JsonRpcError,
    ErrorCode,
    PROTOCOL_VERSION
)
from .registry import ToolRegistry

class ExecutionBridge:
    """Bridge for executing tools across languages"""
    
    def __init__(self):
        self.typescript_process: Optional[subprocess.Popen] = None
        self.request_id = 0
        self.pending_requests: Dict[str, asyncio.Future] = {}
        self.running = False
    
    async def initialize(self, typescript_path: str = 'node') -> None:
        """Initialize the TypeScript bridge"""
        if self.typescript_process:
            return
        
        # Start TypeScript bridge process
        bridge_script = Path(__file__).parent.parent / 'typescript' / 'dist' / 'bridge.js'
        self.typescript_process = subprocess.Popen(
            [typescript_path, str(bridge_script), '--mode', 'client'],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            bufsize=1
        )
        
        self.running = True
        
        # Start reader task
        asyncio.create_task(self._read_typescript_output())
        
        # Wait for ready
        await self._wait_for_ready()
    
    async def execute(
        self,
        tool_id: str,
        language: str,
        parameters: Any,
        context: ToolContext
    ) -> ToolResult:
        """Execute a tool"""
        if language == 'python':
            return await self._execute_python(tool_id, parameters, context)
        else:
            return await self._execute_typescript(tool_id, parameters, context)
    
    async def _execute_python(
        self,
        tool_id: str,
        parameters: Any,
        context: ToolContext
    ) -> ToolResult:
        """Execute a Python tool"""
        tool = ToolRegistry.get(tool_id, 'python')
        if not tool:
            raise ValueError(f"Python tool '{tool_id}' not found")
        
        # Validate parameters
        valid, errors = ToolRegistry.validate_parameters(tool_id, 'python', parameters)
        if not valid:
            raise ValueError(f"Invalid parameters: {', '.join(errors)}")
        
        # Execute with timeout
        try:
            result = await asyncio.wait_for(
                tool.handler(parameters, context),
                timeout=context.timeout
            )
            return result
        except asyncio.TimeoutError:
            raise TimeoutError(f"Tool execution timed out after {context.timeout}s")
    
    async def _execute_typescript(
        self,
        tool_id: str,
        parameters: Any,
        context: ToolContext
    ) -> ToolResult:
        """Execute a TypeScript tool via bridge"""
        if not self.typescript_process:
            await self.initialize()
        
        request_id = str(self.request_id)
        self.request_id += 1
        
        request = {
            'jsonrpc': '2.0',
            'id': request_id,
            'method': 'tool.execute',
            'protocol': PROTOCOL_VERSION,
            'params': {
                'tool': tool_id,
                'language': 'typescript',
                'parameters': parameters,
                'context': {
                    'sessionId': context.session_id,
                    'messageId': context.message_id,
                    'timeout': int(context.timeout * 1000)  # Convert to milliseconds
                }
            }
        }
        
        # Create future for response
        future = asyncio.Future()
        self.pending_requests[request_id] = future
        
        # Send request
        self.typescript_process.stdin.write(json.dumps(request) + '\n')
        self.typescript_process.stdin.flush()
        
        # Wait for response with timeout
        try:
            response = await asyncio.wait_for(future, timeout=context.timeout)
            if 'error' in response:
                raise ValueError(response['error']['message'])
            return ToolResult(**response['result'])
        except asyncio.TimeoutError:
            self.pending_requests.pop(request_id, None)
            raise TimeoutError(f"TypeScript tool execution timed out after {context.timeout}s")
    
    async def _read_typescript_output(self) -> None:
        """Read output from TypeScript process"""
        while self.running and self.typescript_process:
            try:
                line = await asyncio.get_event_loop().run_in_executor(
                    None, self.typescript_process.stdout.readline
                )
                if not line:
                    break
                
                line = line.strip()
                if not line:
                    continue
                
                try:
                    response = json.loads(line)
                    self._handle_response(response)
                except json.JSONDecodeError:
                    print(f"Failed to parse TypeScript response: {line}", file=sys.stderr)
            except Exception as e:
                print(f"Error reading TypeScript output: {e}", file=sys.stderr)
                break
    
    def _handle_response(self, response: Dict[str, Any]) -> None:
        """Handle response from TypeScript"""
        request_id = str(response.get('id', ''))
        future = self.pending_requests.pop(request_id, None)
        
        if future and not future.done():
            future.set_result(response)
    
    async def _wait_for_ready(self) -> None:
        """Wait for TypeScript bridge to be ready"""
        for _ in range(100):  # 10 second timeout
            request = {
                'jsonrpc': '2.0',
                'id': 'ready-check',
                'method': 'ping',
                'protocol': PROTOCOL_VERSION
            }
            
            future = asyncio.Future()
            self.pending_requests['ready-check'] = future
            
            self.typescript_process.stdin.write(json.dumps(request) + '\n')
            self.typescript_process.stdin.flush()
            
            try:
                response = await asyncio.wait_for(future, timeout=0.1)
                if response.get('result') == 'pong':
                    return
            except asyncio.TimeoutError:
                pass
            
            await asyncio.sleep(0.1)
        
        raise RuntimeError("TypeScript bridge failed to start")
    
    async def shutdown(self) -> None:
        """Shutdown the bridge"""
        self.running = False
        
        if self.typescript_process:
            # Send shutdown request
            request = {
                'jsonrpc': '2.0',
                'id': 'shutdown',
                'method': 'shutdown',
                'protocol': PROTOCOL_VERSION
            }
            
            self.typescript_process.stdin.write(json.dumps(request) + '\n')
            self.typescript_process.stdin.flush()
            
            # Wait for process to exit
            try:
                await asyncio.wait_for(
                    asyncio.get_event_loop().run_in_executor(
                        None, self.typescript_process.wait
                    ),
                    timeout=5.0
                )
            except asyncio.TimeoutError:
                self.typescript_process.terminate()
                await asyncio.sleep(0.5)
                if self.typescript_process.poll() is None:
                    self.typescript_process.kill()
            
            self.typescript_process = None
    
    async def register_typescript_module(self, module_path: str) -> None:
        """Register TypeScript tools from a module"""
        if not self.typescript_process:
            await self.initialize()
        
        request = {
            'jsonrpc': '2.0',
            'id': str(self.request_id),
            'method': 'register.module',
            'protocol': PROTOCOL_VERSION,
            'params': {'module': module_path}
        }
        self.request_id += 1
        
        future = asyncio.Future()
        self.pending_requests[request['id']] = future
        
        self.typescript_process.stdin.write(json.dumps(request) + '\n')
        self.typescript_process.stdin.flush()
        
        try:
            await asyncio.wait_for(future, timeout=5.0)
        except asyncio.TimeoutError:
            self.pending_requests.pop(request['id'], None)
            raise TimeoutError("Registration timeout")

# Global bridge instance
_bridge = ExecutionBridge()

async def get_bridge() -> ExecutionBridge:
    """Get the global bridge instance"""
    return _bridge

File: protocol/typescript/src/bridge.ts
import { spawn } from 'child_process'
import { 
  ToolContext, 
  ToolResult, 
  JsonRpcRequest, 
  JsonRpcResponse,
  ErrorCode,
  ToolExecuteRequest,
  PROTOCOL_VERSION
} from './types'
import { ToolRegistry } from './registry'

export class ExecutionBridge {
  private static pythonProcess: any = null
  private static requestId = 0
  private static pendingRequests = new Map<string, {
    resolve: (result: any) => void
    reject: (error: any) => void
    timeout: NodeJS.Timeout
  }>()

  /**
   * Initialize the Python bridge
   */
  static async initialize(pythonPath = 'python'): Promise<void> {
    if (this.pythonProcess) {
      return
    }

    // Start Python bridge process
    this.pythonProcess = spawn(pythonPath, [
      '-m', 
      'protocol.python.bridge',
      '--mode', 'server'
    ], {
      stdio: ['pipe', 'pipe', 'pipe']
    })

    // Handle incoming messages
    this.pythonProcess.stdout.on('data', (data: Buffer) => {
      const lines = data.toString().split('\n').filter(line => line.trim())
      for (const line of lines) {
        try {
          const response: JsonRpcResponse = JSON.parse(line)
          this.handleResponse(response)
        } catch (error) {
          console.error('Failed to parse Python response:', error)
        }
      }
    })

    // Handle errors
    this.pythonProcess.stderr.on('data', (data: Buffer) => {
      console.error('Python bridge error:', data.toString())
    })

    // Handle process exit
    this.pythonProcess.on('exit', (code: number) => {
      console.error(`Python bridge exited with code ${code}`)
      this.pythonProcess = null
      // Reject all pending requests
      for (const [id, pending] of this.pendingRequests) {
        clearTimeout(pending.timeout)
        pending.reject(new Error('Python bridge process exited'))
      }
      this.pendingRequests.clear()
    })

    // Wait for ready signal
    await this.waitForReady()
  }

  /**
   * Execute a tool
   */
  static async execute(
    toolId: string,
    language: 'typescript' | 'python',
    parameters: any,
    context: ToolContext
  ): Promise<ToolResult> {
    if (language === 'typescript') {
      return this.executeTypeScript(toolId, parameters, context)
    } else {
      return this.executePython(toolId, parameters, context)
    }
  }

  /**
   * Execute a TypeScript tool
   */
  private static async executeTypeScript(
    toolId: string,
    parameters: any,
    context: ToolContext
  ): Promise<ToolResult> {
    const tool = ToolRegistry.get(toolId, 'typescript')
    if (!tool) {
      throw new Error(`TypeScript tool '${toolId}' not found`)
    }

    // Validate parameters
    const validation = await ToolRegistry.validateParameters(toolId, 'typescript', parameters)
    if (!validation.valid) {
      throw new Error(`Invalid parameters: ${validation.errors?.join(', ')}`)
    }

    // Execute with timeout
    const timeoutPromise = new Promise<never>((_, reject) => {
      setTimeout(() => reject(new Error('Tool execution timeout')), context.timeout)
    })

    const executionPromise = tool.handler(parameters, context)

    try {
      return await Promise.race([executionPromise, timeoutPromise])
    } catch (error) {
      if (error instanceof Error && error.message === 'Tool execution timeout') {
        throw error
      }
      throw error
    }
  }

  /**
   * Execute a Python tool
   */
  private static async executePython(
    toolId: string,
    parameters: any,
    context: ToolContext
  ): Promise<ToolResult> {
    if (!this.pythonProcess) {
      await this.initialize()
    }

    const requestId = String(++this.requestId)
    
    const request: ToolExecuteRequest = {
      jsonrpc: '2.0',
      id: requestId,
      method: 'tool.execute',
      protocol: PROTOCOL_VERSION,
      params: {
        tool: toolId,
        language: 'python',
        parameters,
        context: {
          sessionId: context.sessionId,
          messageId: context.messageId,
          timeout: context.timeout
        }
      }
    }

    return new Promise((resolve, reject) => {
      // Set up timeout
      const timeout = setTimeout(() => {
        this.pendingRequests.delete(requestId)
        reject(new Error('Python tool execution timeout'))
      }, context.timeout)

      // Store pending request
      this.pendingRequests.set(requestId, { resolve, reject, timeout })

      // Send request
      this.pythonProcess.stdin.write(JSON.stringify(request) + '\n')
    })
  }

  /**
   * Handle response from Python
   */
  private static handleResponse(response: JsonRpcResponse): void {
    const pending = this.pendingRequests.get(String(response.id))
    if (!pending) {
      console.warn('Received response for unknown request:', response.id)
      return
    }

    this.pendingRequests.delete(String(response.id))
    clearTimeout(pending.timeout)

    if (response.error) {
      pending.reject(new Error(response.error.message))
    } else {
      pending.resolve(response.result)
    }
  }

  /**
   * Wait for Python bridge to be ready
   */
  private static async waitForReady(): Promise<void> {
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('Python bridge failed to start'))
      }, 10000)

      const checkReady = () => {
        const request: JsonRpcRequest = {
          jsonrpc: '2.0',
          id: 'ready-check',
          method: 'ping',
          protocol: PROTOCOL_VERSION
        }

        this.pythonProcess.stdin.write(JSON.stringify(request) + '\n')

        const listener = (data: Buffer) => {
          try {
            const response: JsonRpcResponse = JSON.parse(data.toString())
            if (response.id === 'ready-check' && response.result === 'pong') {
              clearTimeout(timeout)
              this.pythonProcess.stdout.removeListener('data', listener)
              resolve()
            }
          } catch (error) {
            // Ignore parse errors during startup
          }
        }

        this.pythonProcess.stdout.on('data', listener)
      }

      setTimeout(checkReady, 100)
    })
  }

  /**
   * Shutdown the bridge
   */
  static async shutdown(): Promise<void> {
    if (!this.pythonProcess) {
      return
    }

    // Send shutdown request
    const request: JsonRpcRequest = {
      jsonrpc: '2.0',
      id: 'shutdown',
      method: 'shutdown',
      protocol: PROTOCOL_VERSION
    }

    this.pythonProcess.stdin.write(JSON.stringify(request) + '\n')

    // Wait for process to exit
    await new Promise<void>((resolve) => {
      this.pythonProcess.on('exit', () => resolve())
      setTimeout(() => {
        this.pythonProcess.kill()
        resolve()
      }, 5000)
    })

    this.pythonProcess = null
  }

  /**
   * Register Python tools from a module
   */
  static async registerPythonModule(modulePath: string): Promise<void> {
    if (!this.pythonProcess) {
      await this.initialize()
    }

    const request: JsonRpcRequest = {
      jsonrpc: '2.0',
      id: String(++this.requestId),
      method: 'register.module',
      protocol: PROTOCOL_VERSION,
      params: { module: modulePath }
    }

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pendingRequests.delete(request.id as string)
        reject(new Error('Registration timeout'))
      }, 5000)

      this.pendingRequests.set(request.id as string, {
        resolve: () => {
          clearTimeout(timeout)
          resolve()
        },
        reject: (error) => {
          clearTimeout(timeout)
          reject(error)
        },
        timeout
      })

      this.pythonProcess.stdin.write(JSON.stringify(request) + '\n')
    })
  }
}